### PR TITLE
Add S3 Http Proxy configuration support

### PR DESCRIPTION
**Pull request by aczajkowski**
## Description

Add S3 Http Proxy configuration support for Trino

## Documentation

(X) No documentation is needed.
(V) Sufficient documentation is included in this PR.
(X) Documentation PR is available with #prnumber.
(X) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

(X) No release notes entries required.
(V) Release notes entries required with the following suggested text:

```markdown
# Hive/Iceberg
* Support S3 Http Proxy configuration for Trino
```

### CODE REVIEW

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/HiveS3Config.java [hunk 1]
```diff
@@ -493,4 +502,88 @@ public HiveS3Config setS3StreamingPartSize(DataSize s3StreamingPartSize)
...
+
+    public List<String> getS3NonProxyHosts()
+    {
+        return s3nonProxyHosts;
+    }
+
+    @Config("hive.s3.proxy.non-proxy-hosts")
+    public HiveS3Config setS3NonProxyHosts(List<String> s3nonProxyHosts)
+    {
+        this.s3nonProxyHosts = s3nonProxyHosts;
```
Comment (losipiuk) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/HiveS3Config.java:550]: ImmutableList.copyOf

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/s3/TestHiveS3Config.java [hunk 2]
```diff
@@ -110,6 +118,13 @@ public void testExplicitPropertyMappings()
                 .put("hive.s3.requester-pays.enabled", "true")
                 .put("hive.s3.streaming.enabled", "false")
                 .put("hive.s3.streaming.part-size", "15MB")
+                .put("hive.s3.proxy.host", "localhost")
+                .put("hive.s3.proxy.port", "14000")
+                .put("hive.s3.proxy.ssl.enabled", "false")
+                .put("hive.s3.proxy.non-proxy-hosts", "test")
```
Comment (losipiuk) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/s3/TestHiveS3Config.java:124]: test with longer list

File: docs/src/main/sphinx/connector/hive-s3.rst [hunk 3]
```diff
@@ -92,6 +92,22 @@ Property Name                                Description
                                              without staging file to be created in the local file system.
 
 ``hive.s3.streaming.part-size``              The part size for S3 streaming upload. Defaults to ``16MB``.
+
+``hive.s3.proxy.host``                       Proxy host to use if connecting through a proxy
+
+``hive.s3.proxy.port``                       Proxy port to use if connecting through a proxy
+
+``hive.s3.proxy.ssl.enabled``                Use HTTPS to communicate with Proxy, defaults to ``true``.
```
Comment (losipiuk) [inline docs/src/main/sphinx/connector/hive-s3.rst:100]: It feels more like "should https:// request be handled by the proxy".
Maybe keep the nomenclature AWS uses (hive.s3.proxy.protocol)

**Review by losipiuk — APPROVED** [review]
LGTM

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/HiveS3Config.java [hunk 4]
```diff
@@ -68,6 +71,13 @@
     private boolean requesterPaysEnabled;
     private boolean s3StreamingUploadEnabled = true;
     private DataSize s3StreamingPartSize = DataSize.of(16, MEGABYTE);
+    private String s3proxyHost;
+    private Integer s3proxyPort = -1;
+    private Protocol s3ProxyProtocol = Protocol.HTTPS;
```
Comment (losipiuk) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/HiveS3Config.java:76]: Better replicate enum definition in Trino. So we have control over value set.

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/HiveS3Config.java [hunk 5]
```diff
@@ -493,4 +503,88 @@ public HiveS3Config setS3StreamingPartSize(DataSize s3StreamingPartSize)
...
+
+    public Protocol getS3ProxyProtocol()
+    {
+        return s3ProxyProtocol;
+    }
+
+    @Config("hive.s3.proxy.protocol")
+    public HiveS3Config setS3ProxyProtocol(String s3ProxyProtocol)
+    {
+        this.s3ProxyProtocol = Protocol.valueOf(s3ProxyProtocol);
```
Comment (losipiuk) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/HiveS3Config.java:539]: Allow for passing lowercase values.

**Review by losipiuk — APPROVED** [review]
LGTM. Is there a way to test it? Maybe an integration test with MinIO + a HTTP proxy container (not sure if worth it).

Also did you have a chance to at least test it manually?

**Comment by aczajkowski** [comment]
> LGTM. Is there a way to test it? Maybe an integration test with MinIO + a HTTP proxy container (not sure if worth it).
> 
> Also did you have a chance to at least test it manually?

@losipiuk 
No did not have enough time yet to test it manually. 
I was thinking on doing such Integration test with MinIO and HTTP Proxy. 
But i have doubts:
- its going to be slow and memory consuming (and we are already on the edge of performance of our CI)
- seems we will be mostly testing AWS client library code.

On our site we test if configuration is properly handled. But not now TrinoS3FileSystem is configured. 
Possibly i could even should add a test for that. This would cover our code well and check if config settings are properly set to S3 Client through all layers.
WDYT ??

**Comment by losipiuk** [comment]
> > LGTM. Is there a way to test it? Maybe an integration test with MinIO + a HTTP proxy container (not sure if worth it).
> > Also did you have a chance to at least test it manually?
> 
> @losipiuk No did not have enough time yet to test it manually. I was thinking on doing such Integration test with MinIO and HTTP Proxy. But i have doubts:
> 
> * its going to be slow and memory consuming (and we are already on the edge of performance of our CI)
> * seems we will be mostly testing AWS client library code.

Why do you think it would be slow. It would as fast as any other integration test where we put up MinIO container. It may be a bit painful to code - but runtime-wise I do not expect it will be a problem.

> 
> On our site we test if configuration is properly handled. But not how TrinoS3FileSystem is configured. Possibly i could even should add a test for that. This would cover our code well and check if config settings are properly set to S3 Client through all layers. WDYT ??

You  mean tests like we have in `TestTrinoS3FileSystem`. Would be nice to have.

**Comment by aczajkowski** [comment]
> > @losipiuk No did not have enough time yet to test it manually. I was thinking on doing such Integration test with MinIO and HTTP Proxy. But i have doubts:
> > 
> > * its going to be slow and memory consuming (and we are already on the edge of performance of our CI)
> > * seems we will be mostly testing AWS client library code.
> 
> Why do you think it would be slow. It would as fast as any other integration test where we put up MinIO container. It may be a bit painful to code - but runtime-wise I do not expect it will be a problem.

From my observations they are slower. But yes it's not going to be supper heavy.
Maybe let me rephrase my doubt. I think it's going to mostly test S3 Client from AWS.
Which i belive is well tests by AWS.
 
> > On our site we test if configuration is properly handled. But not how TrinoS3FileSystem is configured. Possibly i could even should add a test for that. This would cover our code well and check if config settings are properly set to S3 Client through all layers. WDYT ??
> 
> You mean tests like we have in `TestTrinoS3FileSystem`. Would be nice to have.

Yes just adde it :)

**Comment by aczajkowski** [comment]
@losipiuk I've adde Integration test using Apache HTTPD + MinIO to test S3 Proxy configuration.

**Comment by aczajkowski** [comment]
@losipiuk reverted to single commit (all changes connected to Proxy Integration Test dropped)

**Review by losipiuk — APPROVED** [review]

**Comment by losipiuk** [comment]
CI: https://github.com/trinodb/trino/issues/4936

**Comment by findepi** [comment]
test coming here: https://github.com/trinodb/trino/pull/11310
